packages/permission-common/src/PermissionClient.ts

```diff
@@ -0,0 +1,122 @@
+import fetch from 'cross-fetch';
+import * as uuid from 'uuid';
+import {
+  AuthorizeResult,
+  AuthorizeRequest,
+  AuthorizeResponse,
+  Identified,
+} from './types/api';
+import { DiscoveryApi } from './types/discovery';
+
+/**
+ * Options for authorization requests; currently only an optional auth token.
+ * @public
+ */
+export type AuthorizeRequestOptions = {
+  token?: string;
+};
+
+/**
+ * An isomorphic client for requesting authorization for Backstage permissions.
+ * @public
+ */
+export class PermissionClient {
+  private readonly discoveryApi: DiscoveryApi;
+
+  constructor(options: { discoveryApi: DiscoveryApi }) {
+    this.discoveryApi = options.discoveryApi;
+  }
+
+  /**
+   * Request authorization from the permission-backend for the given set of permissions.
+   *
+   * Authorization requests check that a given Backstage user can perform a protected operation,
+   * potentially for a specific resource (such as a catalog entity). The Backstage identity token
+   * should be included in the `options` if available.
+   *
+   * Permissions can be imported from plugins exposing them, such as `catalogEntityReadPermission`.
+   *
+   * The response will be either ALLOW or DENY when either the permission has no resourceType, or a
+   * resourceRef is provided in the request. For permissions with a resourceType, CONDITIONAL may be
+   * returned if no resourceRef is provided in the request. Conditional responses are intended only
+   * for backends which have access to the data source for permissioned resources, so that filters
+   * can be applied when loading collections of resources.
+   * @public
+   */
+  async authorize(
+    requests: AuthorizeRequest[],
+    options?: AuthorizeRequestOptions,
+  ): Promise<AuthorizeResponse[]> {
+    const identifiedRequests: Identified<AuthorizeRequest>[] = requests.map(
+      request => ({
+        id: uuid.v4(),
+        ...request,
+      }),
+    );
+
+    const permissionApi = await this.discoveryApi.getBaseUrl('permission');
+    const response = await fetch(`${permissionApi}/authorize`, {
+      method: 'POST',
+      body: JSON.stringify(identifiedRequests),
+      headers: {
+        ...this.getAuthorizationHeader(options?.token),
+        'content-type': 'application/json',
+      },
+    });
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+
+    const identifiedResponses = await response.json();
+    this.assertValidResponses(identifiedRequests, identifiedResponses);
+
+    const responsesById = identifiedResponses.reduce((acc, r) => {
+      acc[r.id] = r;
+      return acc;
+    }, {} as Record<string, Identified<AuthorizeResponse>>);
+
+    return identifiedRequests.map(request => responsesById[request.id]);
+  }
+
+  private getAuthorizationHeader(token?: string): Record<string, string> {
+    return token ? { Authorization: `Bearer ${token}` } : {};
+  }
+
+  private assertValidResponses(
```

Comment:
Kudos for doing response validation, however it feels like this could be simplified by using a library like https://github.com/colinhacks/zod . We have discussed using it more as a replacement for yup and this might be a great use case. Freben already used it once for the backend-tasks package.

Reply:
Thanks for the suggestion @jhaals! Pushed d47fc24c2d014c249701378dc3d2085d1ee6ffc3 for this. I played around with using Zod to define the original types, but it ends up leaking out to the api-report so I opted for including both the schema and the corresponding types in the end. Interested to hear feedback on this - we can reduce duplication by using Zod to define the shared types for things like AuthorizeResponse and PermissionCriteria, with the drawbacks that a) the types are much less readable looking at the code only, and b) the schemas need to leak into the api-report, since the public types are inferred from them.

Reply:
Cool! I think it makes sense to do it the way you did with duplication in order to get nice types and avoid leaking of zod into the api report. I'd probably always vote for readability over duplication.
I don't know if you have anything to add @freben from previous experience using the library otherwise I think that's the way to go.